README.md

```diff
@@ -1,12 +1,332 @@
+The most common linear operators are (potentially structured) matrices,
+where the function applying them to a vector are (potentially efficient)
+matrix-vector multiplication routines.
+
+In code, a `LinearOperator` is a class that
+1. specifies the tensor(s) needed to define the LinearOperator,
+1. specifies a `_matmul` function (how the LinearOperator is applied to a vector),
+1. specifies a `_size` function (how big is the LinearOperator if it is represented as a matrix, or batch of matrices), and
+1. specifies a `_transpose_nonbatch` function (the adjoint of the LinearOperator).
+
+For example:
+
+```python
+class DiagLinearOperator(linear_operator.LinearOperator):
+    r"""
+    A LinearOperator representing a diagonal matrix.
+    """
+    def __init__(self, diag):
+        # diag: the vector that defines the diagonal of the matrix
+        self.diag = diag
+
+    def _matmul(self):
+        return torch.Size([*self.diag.shape, self.diag.size(-1)])
+
+    def _size(self):
+        return torch.Size([*self.diag.shape, self.diag.size(-1)])
+
+    def _transpose_nonbatch(self):
+        return self  # Diagonal matrices are symmetric
+
+
+# ...
+
+D = DiagLinearOperator(torch.tensor([1., 2., 3.])
+# Represents the matrix
+#   [[1., 0., 0.],
+#    [0., 2., 0.],
+#    [0., 0., 3.]]
+torch.matmul(D, torch.tensor([4., 5., 6.])
+# Returns [4., 10., 18.]
+```
+
+Other functions (e.g. `logdet`, `eigh`, etc.) can be defined as needed if efficient routines exist.
+
+#### Why is This Useful?
+While `_matmul`, `_size`, and `_transpose_nonbatch` might seem like a limited set of functions,
+it turns out that most functions on the `torch` and `torch.linalg` namespaces can be efficiently implemented
+using only these three primitative functions.
+
+Moreover, because `_matmul` is a linear function, it is very easy to compose linear operators in various ways.
+For example: adding two linear operators (`SumLinearOperator`) just requires adding the output of their `_matmul` functions.
+This makes it possible to define very complex compositional structures that still yield efficient linear algebraic routines.
+
+
+## Use Cases
+
+There are several use cases for the LinearOperator package.
+Here we highlight two general themes:
+
+### Modular Code for Structured Matrices
+
+For example, let's say that you have a generative model that involves
+sampling from a high-dimensional multivariate Gaussian.
+This sampling operation will require storing and manipulating a large covariance matrix,
+so to speed things up you might want to experiment with different structured
+approximations of that covariance matrix.
+This is easy with the LinearOperator package.
+
+```python
+from gpytorch.distributions import MultivariateNormal
+
+# variance = torch.randn(10000)
+cov = DiagLinearOperator(variance)
+# or
+# cov = RootLinearOperator(...) + DiagLinearOperator(...)
+# or
+# cov = KroneckerProductLinearOperator(...)
+# or
+# cov = ToeplitzLinearOperator(...)
```

Comment:
So I played around with this, and unfortunately this doesn't work as well as we'd like (yet). E.g. the low rank + diag case will in the `MultivariateNormal` code still perform a Cholesky decomposition, which isn't exploiting the structure.

Reply:
Whoops - if this becomes `LowRankRootLinearOperator`, then it uses Woodbury. I'm going to fix that now.